logstash-core/src/main/java/org/logstash/secret/store/backend/JavaKeyStore.java

```diff
@@ -0,0 +1,215 @@
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.nio.file.attribute.PosixFileAttributeView;
+import java.nio.file.attribute.PosixFilePermissions;
+import java.security.KeyStore;
+import java.security.KeyStore.PasswordProtection;
+import java.security.KeyStore.ProtectionParameter;
+import java.security.KeyStoreException;
+import java.security.UnrecoverableKeyException;
+import java.util.*;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * <p>Java Key Store implementation for the {@link SecretStore}.</p>
+ * <p>Note this implementation should not be used for high volume or large datasets.</p>
+ * <p>This class is threadsafe.</p>
+ */
+public final class JavaKeyStore implements SecretStore {
+    public static final String LOGSTASH_MARKER = "logstash-key-store";
+    private static final Logger LOGGER = LogManager.getLogger(JavaKeyStore.class);
+    private static final String KEYSTORE_TYPE = "pkcs12";
+
+    private final char[] keyStorePass;
+    private final Path keyStorePath;
+    private final ProtectionParameter protectionParameter;
+    private final Lock readLock;
+    private final Lock writeLock;
+    private KeyStore keyStore;
+
+    /**
+     * Constructor - will create the keystore if it does not exist
+     *
+     * @param keyStorePath The full path to the java keystore
+     * @param keyStorePass The password to the keystore
+     * @throws SecretStoreException if errors occur while trying to create or access the keystore
+     */
+    public JavaKeyStore(Path keyStorePath, char[] keyStorePass) {
+        try {
+            this.keyStorePath = keyStorePath;
+            this.keyStore = KeyStore.getInstance(KEYSTORE_TYPE);
+            this.keyStorePass = SecretStoreUtil.base64Encode(keyStorePass);
+            ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+            readLock = readWriteLock.readLock();
+            writeLock = readWriteLock.writeLock();
+            this.protectionParameter = new PasswordProtection(this.keyStorePass);
+            SecretIdentifier logstashMarker = new SecretIdentifier(LOGSTASH_MARKER);
+
+            try (final InputStream is = Files.newInputStream(keyStorePath)) {
+                try {
+                    keyStore.load(is, this.keyStorePass);
+                } catch (IOException ioe) {
+                    if (ioe.getCause() instanceof UnrecoverableKeyException) {
+                        throw new SecretStoreException.AccessException(
+                                String.format("Can not access Java keystore at %s, check file permissions and the keystore password", keyStorePath.toAbsolutePath
```

Comment:
period, not comma `at %s` recommend also `Please verify correct file permissions and keystore password`

Reply:
will do, thanks.